File: src/lib/mdns.js
const bonjourInstance = require('bonjour')();

function advertise(port, props = {}, log = console) {
  const serviceConfig = {
    name: props.name || 'Unified Hi-Fi Control',
    type: 'roonknob',
    protocol: 'tcp',
    port,
    txt: {
      base: props.base || `http://localhost:${port}`,
      api: '1',
      ...props.txt
    }
  };

  log.info('mDNS: Publishing service', {
    name: serviceConfig.name,
    type: `_${serviceConfig.type}._${serviceConfig.protocol}`,
    port: serviceConfig.port,
    txt: serviceConfig.txt
  });

  const service = bonjourInstance.publish(serviceConfig);

  service.on('up', () => {
    log.info('mDNS: Service published successfully', { name: serviceConfig.name });
  });

  service.on('error', (err) => {
    log.error('mDNS: Service error', { error: err.message || err });
  });

  process.on('exit', () => service.stop());
  return service;
}

/**
 * Discover services on the network via mDNS
 * @param {string} serviceType - Service type to discover (e.g., 'hqplayer' for _hqplayer._tcp)
 * @param {object} opts - Options: timeout (ms, default 5000), log (logger)
 * @returns {Promise<Array>} Array of discovered services
 */
function discover(serviceType, opts = {}) {
  const log = opts.log || console;
  const timeout = opts.timeout || 5000;

  return new Promise((resolve) => {
    const services = [];

    log.info('mDNS: Starting discovery', {
      type: `_${serviceType}._tcp`,
      timeout
    });

    const browser = bonjourInstance.find({ type: serviceType });

    browser.on('up', (service) => {
      log.info('mDNS: Service found', {
        name: service.name,
        host: service.host,
        port: service.port,
        addresses: service.addresses
      });

      services.push({
        name: service.name,
        host: service.host,
        port: service.port,
        addresses: service.addresses || [],
        txt: service.txt || {}
      });
    });

    // Stop after timeout and return results
    setTimeout(() => {
      browser.stop();
      log.info('mDNS: Discovery complete', { found: services.length });
      resolve(services);
    }, timeout);
  });
}

module.exports = { advertise, discover };

File: src/mdns.rs
//! mDNS service advertising for knob discovery
//!
//! Publishes a _roonknob._tcp service so S3 Knob devices can discover the server.

use mdns_sd::{ServiceDaemon, ServiceInfo};
use std::collections::HashMap;

/// Advertise the service via mDNS
pub fn advertise(port: u16, name: &str, base_url: &str) -> anyhow::Result<ServiceDaemon> {
    let mdns = ServiceDaemon::new()?;

    // Build TXT records
    let mut txt = HashMap::new();
    txt.insert("base".to_string(), base_url.to_string());
    txt.insert("api".to_string(), "1".to_string());

    // Create service info
    // Type is "_roonknob._tcp.local."
    let service_type = "_roonknob._tcp.local.";

    // Get hostname and ensure it ends with ".local." for mdns_sd
    let raw_hostname = gethostname::gethostname().to_string_lossy().to_string();
    let hostname = if raw_hostname.ends_with(".local.") {
        raw_hostname
    } else if raw_hostname.ends_with(".local") {
        format!("{}.", raw_hostname)
    } else {
        format!("{}.local.", raw_hostname)
    };

    let service_info = ServiceInfo::new(
        service_type,
        name,
        &hostname,
        (), // Will be filled by enable_addr_auto()
        port,
        Some(txt),
    )?
    .enable_addr_auto();

    tracing::info!(
        "mDNS: Publishing service '{}' on port {} (type: {})",
        name,
        port,
        service_type
    );

    // Register the service
    mdns.register(service_info)?;

    tracing::info!("mDNS: Service registered successfully");

    Ok(mdns)
}
